src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -2079,65 +1771,41 @@ JRT_LEAF(void, SharedRuntime::fixup_callers_callsite(Method* method, address cal
-  // to patch. If we hit the window where the callee has a c2i in the
-  // from_compiled_entry and the null isn't present yet then we lose the race
-  // and patch the code with the same old data. Asi es la vida.
-
-  if (moop->code() == nullptr) return;
-
-  if (nm->is_in_use()) {
-    // Expect to find a native call there (unless it was no-inline cache vtable dispatch)
-    CompiledICLocker ic_locker(nm);
-    if (NativeCall::is_call_before(return_pc)) {
-      ResourceMark mark;
-      NativeCallWrapper* call = nm->call_wrapper_before(return_pc);
-      //
-      // bug 6281185. We might get here after resolving a call site to a vanilla
-      // virtual call. Because the resolvee uses the verified entry it may then
-      // see compiled code and attempt to patch the site by calling us. This would
-      // then incorrectly convert the call site to optimized and its downhill from
-      // there. If you're lucky you'll get the assert in the bugid, if not you've
-      // just made a call site that could be megamorphic into a monomorphic site
-      // for the rest of its life! Just another racing bug in the life of
-      // fixup_callers_callsite ...
-      //
-      RelocIterator iter(nm, call->instruction_address(), call->next_instruction_address());
-      iter.next();
-      assert(iter.has_current(), "must have a reloc at java call site");
-      relocInfo::relocType typ = iter.reloc()->type();
-      if (typ != relocInfo::static_call_type &&
-           typ != relocInfo::opt_virtual_call_type &&
-           typ != relocInfo::static_stub_type) {
-        return;
-      }
-      if (nm->method()->is_continuation_enter_intrinsic()) {
-        if (ContinuationEntry::is_interpreted_call(call->instruction_address())) {
-          return;
-        }
-      }
-      address destination = call->destination();
-      address entry_point = callee->verified_entry_point();
-      if (should_fixup_call_destination(destination, entry_point, caller_pc, moop, cb)) {
-        call->set_destination_mt_safe(entry_point);
-      }
-    }
+  if (!caller->is_in_use() || !NativeCall::is_call_before(return_pc)) {
+    return;
+  }
+
+  // Expect to find a native call there (unless it was no-inline cache vtable dispatch)
+  CompiledICLocker ic_locker(caller);
+  ResourceMark rm;
+
+  // If we got here through a static call or opt_virtual call, then we know where the
+  // call address would be; let's peek at it
+  address callsite_addr = (address)nativeCall_before(return_pc);
+  RelocIterator iter(caller, callsite_addr, callsite_addr + 1);
+  if (!iter.next()) {
+    // No reloc entry found; not a static or optimized virtual call
+    return;
   }
+
+  relocInfo::relocType type = iter.reloc()->type();
+  if (type != relocInfo::static_call_type &&
+      type != relocInfo::opt_virtual_call_type) {
+    return;
+  }
+
+  CompiledDirectCall* callsite = CompiledDirectCall::before(return_pc);
+  callsite->set_to_clean();
```

Comment:
Clever way to get rid of should_fixup_call_destination(), something I've wanted to do!